### DISCUSSION TITLE
Can't Compile: hb-ot.h not found

### DISCUSSION THREAD

**Discussion by henrijaeger** [osmandapp/OsmAnd#18597] [Q&A]
Hi all, I am trying to build OsmAnd from source. However, compiling the project always results in a build failure due to a missing header file called "hb-ot.h". 
I already installed harfbuzz (tried both from source as well as using homebrew). I followed the steps on the support page: https://osmand.net/docs/technical/build-osmand/how-to-compile-the-android-version/

Does somebody have a solution for that?

Thanks a lot in advance!

fcntl(): Bad file descriptor
Android NDK: WARNING: APP_PLATFORM android-21 is higher than android:minSdkVersion 1 in ./AndroidManifest.xml. NDK binaries will *not* be compatible with devices older than android-21. See https://android.googlesource.com/platform/ndk/+/master/docs/user/common_problems.md for more information.    
OsmAnd root: jni/../../..
OsmAnd makefiles:  jni/../../../core-legacy/targets/android/OsmAndCore/Android.mk jni/../../../core-legacy/targets/android/expat/Android.mk jni/../../../core-legacy/targets/android/freetype/Android.mk jni/../../../core-legacy/targets/android/giflib/Android.mk jni/../../../core-legacy/targets/android/harfbuzz/Android.mk jni/../../../core-legacy/targets/android/jpeg/Android.mk jni/../../../core-legacy/targets/android/libpng/Android.mk jni/../../../core-legacy/targets/android/lua/Android.mk jni/../../../core-legacy/targets/android/protobuf/Android.mk jni/../../../core-legacy/targets/android/skia/Android.mk
OsmAnd root: jni/../../..
OsmAnd makefiles:  jni/../../../core-legacy/targets/android/OsmAndCore/Android.mk jni/../../../core-legacy/targets/android/expat/Android.mk jni/../../../core-legacy/targets/android/freetype/Android.mk jni/../../../core-legacy/targets/android/giflib/Android.mk jni/../../../core-legacy/targets/android/harfbuzz/Android.mk jni/../../../core-legacy/targets/android/jpeg/Android.mk jni/../../../core-legacy/targets/android/libpng/Android.mk jni/../../../core-legacy/targets/android/lua/Android.mk jni/../../../core-legacy/targets/android/protobuf/Android.mk jni/../../../core-legacy/targets/android/skia/Android.mk
OsmAnd root: jni/../../..
OsmAnd makefiles:  jni/../../../core-legacy/targets/android/OsmAndCore/Android.mk jni/../../../core-legacy/targets/android/expat/Android.mk jni/../../../core-legacy/targets/android/freetype/Android.mk jni/../../../core-legacy/targets/android/giflib/Android.mk jni/../../../core-legacy/targets/android/harfbuzz/Android.mk jni/../../../core-legacy/targets/android/jpeg/Android.mk jni/../../../core-legacy/targets/android/libpng/Android.mk jni/../../../core-legacy/targets/android/lua/Android.mk jni/../../../core-legacy/targets/android/protobuf/Android.mk jni/../../../core-legacy/targets/android/skia/Android.mk
OsmAnd root: jni/../../..
OsmAnd makefiles:  jni/../../../core-legacy/targets/android/OsmAndCore/Android.mk jni/../../../core-legacy/targets/android/expat/Android.mk jni/../../../core-legacy/targets/android/freetype/Android.mk jni/../../../core-legacy/targets/android/giflib/Android.mk jni/../../../core-legacy/targets/android/harfbuzz/Android.mk jni/../../../core-legacy/targets/android/jpeg/Android.mk jni/../../../core-legacy/targets/android/libpng/Android.mk jni/../../../core-legacy/targets/android/lua/Android.mk jni/../../../core-legacy/targets/android/protobuf/Android.mk jni/../../../core-legacy/targets/android/skia/Android.mk
fcntl(): Bad file descriptor
[x86] Compile++      : osmand <= commonRendering.cpp
In file included from jni/../../../core-legacy/targets/android/OsmAndCore/../../../native/src/commonRendering.cpp:1:
jni/../../../core-legacy/targets/android/OsmAndCore/../../../native/src/commonRendering.h:10:10: fatal error: 'hb-ot.h' file not found
#include <hb-ot.h>
         ^~~~~~~~~
1 error generated.
make: *** [/Users/henrijager/Library/Android/sdk/ndk/23.0.7599858/build/core/build-binary.mk:478: obj/local/x86/objs/osmand/__/__/__/native/src/commonRendering.o] Error 1
make: *** Waiting for unfinished jobs....
fcntl(): Bad file descriptor
[x86] Compile++      : osmand <= renderRules.cpp
In file included from jni/../../../core-legacy/targets/android/OsmAndCore/../../../native/src/renderRules.cpp:1:
In file included from jni/../../../core-legacy/targets/android/OsmAndCore/../../../native/src/renderRules.h:18:
jni/../../../core-legacy/targets/android/OsmAndCore/../../../native/src/commonRendering.h:10:10: fatal error: 'hb-ot.h' file not found
#include <hb-ot.h>
         ^~~~~~~~~
1 error generated.
make: *** [/Users/henrijager/Library/Android/sdk/ndk/23.0.7599858/build/core/build-binary.mk:478: obj/local/x86/objs/osmand/__/__/__/native/src/renderRules.o] Error 1

> Task :OsmAnd:buildOsmAndCore FAILED

FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':OsmAnd:buildOsmAndCore'.
> Process 'command 'bash'' finished with non-zero exit value 2

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* What went wrong:
java.lang.StackOverflowError (no error message)

**Comment by vshcherb** [comment]
We use android-ndk-23 to compile, probably it's time to upgrade to android-ndk-26...

**Reply by henrijaeger** [reply]
Thank you for your answer!
I am using Android NDK version 23 since its installed via Android Studio and the environment variable ANDROID_NDK ist set to an Android NDK version of 23.x:
<img width="714" alt="image" src="https://github.com/osmandapp/OsmAnd/assets/64465875/addae826-c456-4fc9-ba78-601021511d93">
<img width="714" alt="image" src="https://github.com/osmandapp/OsmAnd/assets/64465875/2fe8a39e-e9f7-4040-af2d-34693e36cd63">

Do you see any issues with my configuration?

**Reply by vshcherb** [reply]
Not really - though we compile using command line gradle - https://creator.osmand.net:8080/view/OsmAnd%20Builds/job/OsmAndCore-android-ndk23/. You can use precompiled version from ivy repo that what default build does

**Comment by alexsmithbr** [comment]
I had the same issue and fixed it by modifying build-common.gradle like this:

```
diff --git a/OsmAnd/build-common.gradle b/OsmAnd/build-common.gradle
index 573f4fa..7fa1309 100644
--- a/OsmAnd/build-common.gradle
+++ b/OsmAnd/build-common.gradle
@@ -334,7 +334,7 @@ task buildOsmAndCore(type: Exec) {
                if(flavour.length()  > 0) {
                        environment "$flavour", "1"
                }
-               commandLine "bash", file("./old-ndk-build.sh").getAbsolutePath()
+               commandLine "bash", file("./ndk-build.sh").getAbsolutePath()
        } else {
                commandLine "cmd", "/c", "echo", "Not supported"
        }

```

It seems like old-ndk-build.sh brings some incompatibilities...

Hope that helps.

**Reply by alexsmithbr** [reply]
By applying the change I mentioned, I noticed building crashes again. The root cause - I think - is many dependencies are downloaded when you start the building process and, at least for me, some of them were not correctly patched by the automatic patching process inside `core-legacy/externals/skia`. This was due to the fact I didn't have python properly set up... Some patching scripts rely on your system having python available as exactly `python`, not `python3`. For Ubuntu 23.10 (and possibly other versions) users, this package solves the issue:

`sudo apt install python-is-python3`

Then the above patch (changing `old-ndk-build.sh` to `ndk-build.sh` should not be necessary. I'll test this and post the results.

**Reply by alexsmithbr** [reply]
I was able to build successfully with:

```
cd <repo_base_dir>
// removed all files with rm * (careful here)
// re-synch with repo sync
// set up the export variables for android SDK, NDK, etc.
cd android
// add these lines to gradle.properties:
// org.gradle.daemon=true
// org.gradle.jvmargs=-Xmx8192m -Xss1024m
cd OsmAnd
../gradlew assembleAndroidFullLegacyArm64Debug
```

Building the OpenGL version still crashes saying some methods weren't found. At least Legacy version is working...

**Reply by alexsmithbr** [reply]
Being a completely newbie to gradle and friends made me waste tons of hours trying to figure things out... If you're out of ideas why compiling crashes due to methods not found, etc., try removing ~/.gradle directory before starting the building process. I had some fundamental libraries cache-compiled there and they were breaking the build entirely. Removed that dir and restarted the process and whoa-la! OpenGL works too!

```
cd android/OsmAnd
../gradlew assembleAndroidFullOpenglArm64Debug
```
Final hint: if you run into memory problems, make sure you have this line on `<repo_base_dir>/android/gradle.properties`:

```
org.gradle.jvmargs=-Xmx8192m -Xss1024m
```

Change 8192 to suit the amount of memory you want to allow jvm to use.

**Reply by sapstern** [reply]
Thanks Alex, 

your hack switching to "ndk-build.sh" did it for me.

Regards

Matthias
